### PR TITLE
Base.Test: add GenericArray type and use it in test/random.jl

### PR DESCRIPTION
**Pull request by rfourquet**
`GenericArray`, similarly to `GenericString`, `GenericDict` and `GenericSet`, can be used to test APIs using `AbstractArray`.

### CODE REVIEW

File: test/random.jl [hunk 1]
```diff
@@ -370,11 +372,11 @@ for rng in ([], [MersenneTwister(0)], [RandomDevice()])
     for f! in [rand!, randn!, randexp!]
         for T in (f! === rand! ? types : f! === randn! ? cftypes : ftypes)
             X = T == Bool ? T[0,1] : T[0,1,2]
-            for A in (Array{T}(5), Array{T}(2, 3))
+            for A in [Array{T}(5), Array{T}(2, 3), GenericArray{T}(5), GenericArray{T}(2, 3)]
```
Comment (Sacha0) [inline test/random.jl]: Out of curiosity, why an array rather than the former tuple? :)

Comment (rfourquet) [inline test/random.jl]: Oh, just my preference of the moment, sorry for the noise!

**Comment by rfourquet** [comment]
I will merge within a day or two, if no objection.

File: test/random.jl (hunk 1 above)
Comment (tkelman) [inline test/random.jl]: might end up accidentally promoting?

Comment (rfourquet) [inline test/random.jl]: I don't understand... does the `[...]` constructor for arrays sometimes promotes?

Comment (fredrikekre) [inline test/random.jl]: It tries to find a common type before bailing to `Any`
```julia
julia> [1, 1., complex(1,1)]
3-element Array{Complex{Float64},1}:
 1.0+0.0im
 1.0+0.0im
 1.0+1.0im
```

Comment (rfourquet) [inline test/random.jl]: Ah yes sure... so in this case I get `Array{AbstractArray{Int64,1},1}`, and each item keeps its separate type, though I can change it back to a tuple if this is prefered.

Comment (fredrikekre) [inline test/random.jl]: Promotion rules could change, so safer to keep it a tuple.

Comment (rfourquet) [inline test/random.jl]: Ok, will do (but I would feel uncomfortable if this could happen that one array gets promoted to another type in this case...)

**Pull request merged**